src/framework/theme/components/context-menu/context-menu.directive.ts

```diff
@@ -0,0 +1,131 @@
+/**
+ * @license
+ * Copyright Akveo. All Rights Reserved.
+ * Licensed under the MIT License. See License.txt in the project root for license information.
+ */
+
+import { Directive, ElementRef, HostListener, Input, OnDestroy, OnInit } from '@angular/core';
+import { NbPopoverDirective } from '../popover/popover.directive';
+import { NbMenuItem } from '../menu/menu.service';
+import { NbThemeService } from '../../services/theme.service';
+import { NbPopoverAdjustment, NbPopoverPlacement } from '../popover/helpers/model';
+import { NbContextMenuComponent } from './context-menu.component';
+
+/**
+ * Full featured context menu directive.
+ *
+ * ![image](assets/images/components/context-menu.gif)
+ *
+ * @example Juts pass basic menu items:
+ *
+ * ```
+ * <button [nbContextMenu]="items"></button>
+ * ...
+ * items = [{ title: 'Profile' }, { title: 'Log out' }];
+ * ```
+ *
+ * @example Context menu has different placements, such as: top, bottom, left and right
+ * which can be used as following:
+ *
+ * ```
+ * <button [nbContextMenu]="items" nbContextMenuPlacement="right"></button>
+ * ...
+ * items = [{ title: 'Profile' }, { title: 'Log out' }];
+ * ```
+ *
+ * @example By default context menu will try to adjust itself to maximally fit viewport
+ * and provide the best user experience. It will try to change placement of the context menu.
+ * If you wanna disable this behaviour just set it falsy value.
+ *
+ * ```
+ * <button [nbContextMenu]="items" nbContextMenuAdjustment="counterclockwise"></button>
+ * ...
+ * items = [{ title: 'Profile' }, { title: 'Log out' }];
+ * ```
+ * */
+@Directive({ selector: '[nbContextMenu]' })
+export class NbContextMenuDirective implements OnInit, OnDestroy {
+
+  /**
+   * Basic menu items, will be passed to the internal NbMenuComponent.
+   * */
+  @Input('nbContextMenu')
+  set items(items: NbMenuItem[]) {
+    this.validateItems(items);
+    this.popover.context = { items };
+  };
+
+  /**
+   * Position will be calculated relatively host element based on the placement.
+   * Can be top, right, bottom and left.
+   * */
+  @Input('nbContextMenuPlacement')
+  set placement(placement: NbPopoverPlacement) {
+    this.popover.placement = placement;
+  };
+
+  /**
+   * Container placement will be changes automatically based on this strategy if container can't fit view port.
+   * Set this property to any falsy value if you want to disable automatically adjustment.
+   * Available values: clockwise, counterclockwise.
+   * */
+  @Input('nbContextMenuAdjustment')
+  set adjustment(adjustment: NbPopoverAdjustment) {
+    this.popover.adjustment = adjustment;
+  }
+
+  protected popover: NbPopoverDirective;
+
+  constructor(hostRef: ElementRef, themeService: NbThemeService) {
+    /**
+     * Initialize popover with all the important inputs.
+     * */
+    this.popover = new NbPopoverDirective(hostRef, themeService);
```

Comment:
what if we do this differently using `mixins`? I mean we can 'mix' some features into the component class. We better discuss this offline.